### PR TITLE
Disable the screen keyboard if a real keyboard exists

### LINKED ISSUE #692: Gnome Virtual keyboard should be disabled on Microsoft Surface 3 Go laptop when physical keyboard is connected
Microsoft Surface 3 Go laptop has a detachable keyboard. Based on user feedback the virtual/on-screen keyboard should always be disabled when the real keyboard is connected. Virtual keyboard should only appear on the tablet mode when the keyboard is detached.

Some ideas:
- Surface specific hwquik etc. that activates functionality to activate/deactivate on-screen keyboard on this specific hardware
- puavo-conf to actively block on-screen keyboard on any device
- Setting for this to the gnome control panel

https://extensions.gnome.org/extension/3222/block-caribou-36/

### PR DESCRIPTION
**Pull request by tuomasjjrasanen**
This PR introduces a new GNOME Shell Extension: `screenkeyboardcontroller@puavo.org`, which hooks into KeyboardManager and enables/disables screen keyboard in touch mode based on the existence of a real keyboard.

If the device has a real keyboard, screen keyboard is disabled, and vice versa.

In addition, whenever input device setup is changed (Clutter's deviced-added or device-removed signal), `_syncEnabled()` is called to update KeyboardManager's knowledge of it's surroundings.

The relevant code section in KeyboardManager (js/ui/keyboard.js of gnome-shell):

```
    _lastDeviceIsTouchscreen() {
        if (!this._lastDevice)
            return false;

        let deviceType = this._lastDevice.get_device_type();
        return deviceType == Clutter.InputDeviceType.TOUCHSCREEN_DEVICE;
    }

    _syncEnabled() {
        let enableKeyboard = this._a11yApplicationsSettings.get_boolean(SHOW_KEYBOARD);
        let autoEnabled = this._seat.get_touch_mode() && this._lastDeviceIsTouchscreen();
        let enabled = enableKeyboard || autoEnabled;

        if (!enabled && !this._keyboard)
            return;

        if (enabled && !this._keyboard) {
            this._keyboard = new Keyboard();
            this._keyboard.connect('visibility-changed', () => {
                this.emit('visibility-changed');
                this._bottomDragAction.enabled = !this._keyboard.visible;
            });
        } else if (!enabled && this._keyboard) {
            this._keyboard.setCursorLocation(null);
            this._keyboard.destroy();
            this._keyboard = null;
            this._bottomDragAction.enabled = true;
        }
    }
```

So, whenever `_syncEnabled()` is called (it's a callback of multiple different signals) in ["touch mode"](https://mutter.gnome.org/clutter/method.Seat.get_touch_mode.html), our hook function gets to say the last word (except when screen keyboard is enabled from GSettings permanently, but that's a good thing).

`has-real-keyboard` just reads `/proc/bus/input/devices`  and checks whether there's a real keyboard or not; real keyboards are connected to physical bus and have sysrq handlers. If we ever encounter a keyboard which does not fulfill this real requirement, I'm sure we can come up with more complex logic. But currently, reality looks simple, so let's assume it also is simple.

Closes #692

### CODE REVIEW

File: parts/ltsp/client/lib/has-real-keyboard [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+set -eu
+set -o pipefail
+
+while read -a handlers; do
+    for handler in "${handlers[@]}"; do
+        if [ "${handler}" = sysrq ]; then
```
Comment (Mazhoon) [inline parts/ltsp/client/lib/has-real-keyboard]: tested on a couple of devices:
linux smartphone (no keyboard): ok
surface without keyboard: ok
hp laptop: ok
intel nucs, however, seem to have a
`N: Name="ACPI Virtual Keyboard Device"`
always present, which has a handler list of
`H: Handlers=sysrq kbd event4 rfkill`
(checked remotely a couple of devices; didn't physically verify that they don't have a keyboard connected right now, but at least from one I'm pretty confident it does not, and another one had a separate "AT Translated Set 2 keyboard" which is probably what a real connected keyboard looks like. but I can verify that later this week)

**Comment by Mazhoon** [comment]
Tested with a surface without keyboard: no extension, os-keyboard pops up, good. enable extension, os-keyboard pop ups, good. connect bluetooth keyboard, os-keyboard does not pop up, good. disconnect bluetooth keyboard, os-keyboard won't pop up (as last keyboard device was a physical keyboard?). tried tapping, opening menus etc. only after pressing power button, the os-keyboard would pop up again (as power button was considered a not-real keyboard device, I assume)

File: parts/ltsp/client/lib/has-real-keyboard (hunk 1 above)
Comment (tuomasjjrasanen) [reply parts/ltsp/client/lib/has-real-keyboard]: Interesting.. then perhaps another way to filter would be check the bus where the "maybe real keyboard" device is connected to..

**Comment by tuomasjjrasanen** [comment]
> connect bluetooth keyboard, os-keyboard does not pop up, good. disconnect bluetooth keyboard, os-keyboard won't pop up (as last keyboard device was a physical keyboard?). tried tapping, opening menus etc. only after pressing power button, the os-keyboard would pop up again (as power button was considered a not-real keyboard device, I assume)

This has been fixed now, but connecting to device-added and device-removed signals: whenever the input device setup changes, `_syncEnabled()` is called (which is KeyboardManager's function for deducing whether the keyboard should be enabled or not). And because we have replaced `_lastDeviceIsTouchscreen()`, which is called be `_syncEnabled()`, screen keyboard will pop up in those situations (because no hw keyboard is available).

**Comment by tuomasjjrasanen** [comment]
Updated the description.

**Comment by tuomasjjrasanen** [comment]
@juhaerk Integration from puavo-conf `puavo.screenkeyboard.controller.mode` to dconf key `org.gnome.shell.extensions.screenkeyboardcontroller.mode` is left to you.

**Comment by tuomasjjrasanen** [comment]
@juhaerk And now mangling from puavo-conf to dconf is done

File: parts/ltsp/client/lib/has-real-keyboard [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+         for (v in handlers_array) {
+             if (handlers_array[v] == "sysrq") {
+                 printf "%s\n", name
+             }            
+         }
+    }
+}
+' /proc/bus/input/devices)
+
+if [ -z "${real_keyboard_names}" ]; then
```
Comment (Mazhoon) [inline parts/ltsp/client/lib/has-real-keyboard:22]: linux smartphone (no keyboard): ok
surface without keyboard: ok
surface with bt keyboard: ok
hp laptop: ok
some infotv with acpi virtual keyboard (but no physical kb): ok

**Comment by Mazhoon** [comment]
> > connect bluetooth keyboard, os-keyboard does not pop up, good. disconnect bluetooth keyboard, os-keyboard won't pop up (as last keyboard device was a physical keyboard?). tried tapping, opening menus etc. only after pressing power button, the os-keyboard would pop up again (as power button was considered a not-real keyboard device, I assume)
> 
> This has been fixed now, but connecting to device-added and device-removed signals: whenever the input device setup changes, `_syncEnabled()` is called (which is KeyboardManager's function for deducing whether the keyboard should be enabled or not). And because we have replaced `_lastDeviceIsTouchscreen()`, which is called be `_syncEnabled()`, screen keyboard will pop up in those situations (because no hw keyboard is available).

confirmed to work with bluetooth keyboard now: as soon as the system realizes bt kb is turned off, onscreen keyboard pops up -> ok

**Pull request merged**